include/object_pool.h

```diff
@@ -0,0 +1,101 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2024-2024  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_OBJECT_POOL_H
+#define DOSBOX_OBJECT_POOL_H
+
+#include "dosbox.h"
+
+// This class is a generic object re-use pool. Consider using it if your program
+// performs tens of thousands (or more) large object heap allocations as part of
+// its normal and ongoing runtime loop.
+//
+// Ongoing runtime loop:
+//     auto my_obj_ptr = new HugeType(arg1, arg2);
+//     .. do a bit of work ..
+//     delete my_obj_ptr;
+//
+// If you have this pattern, then the ObjectPool is a good candidate:
+//
+//  Instantiate a long-lived pool for your HugeType:
+//     ObjectPool<HugeType> pool = {};
+//
+//  Ongoing runtime loop:
+//     auto my_obj_ptr = pool.Acquire(arg1, arg2);
+//     .. do a bit of work ..
+//     pool.Release(my_obj_ptr);
+
+// Note: this is less about performance and more about fitting a well known
+// use-case (frequent and ongoing heap allocation and destruction of a large
+// object type) with a well known solution (object pool).  The before-and-after
+// benefit will be reduced call outs to the heap and less heap fragmentation.
```

Comment:
Not to nitpick much, but in practice, this is always done due to performance reasons and to avoid memory fragmentation 😄

Reply:
Very nice explanatory comments, btw 🎖️

Reply:
> in practice, this is always done due to performance reasons and to avoid memory fragmentation

lol; good point :-) Will update the wording :+1: 

When I saw the numbers, I was thinking that 4 KB maps re-allocated 10's of thousands of times per second was quite a bit churn (over 4 GB ever 10s or so!), even if we're not leaking them.

I guess even the Raspberry Pi hardware and Linux kernel are pretty efficient at handling that, but figured we might as well avoid that if we can.

Reply:
Updated the wording to mention the performance fragmentation benefits :+1: